Use SDKMan to provision Open JDK 11

pvliss opened the pull request:
Use SDKMAN to provision Open JDK 11 in Azure build

Fixes #9513 

-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

pascalgrimaud commented on azure-pipelines.yml:
```diff
@@ -94,10 +94,11 @@ jobs:
-                    sudo apt-get update
-                    sudo apt-get install -y openjdk-11-jdk
-                    sudo update-java-alternatives -s java-1.11.0-openjdk-amd64
+                    curl -s get.sdkman.io | bash
```
Plz, can you do the same for:
- the Travis file, although it is currently not use
- the Dockerfile

pascalgrimaud replied:
So, I should change this part too, right ?
https://github.com/hipster-labs/jhipster-daily-builds/blob/master/azure-template.yml#L29-L34

pvliss replied:
Updated the travis and docker files as well. 

I guess if you want to check the daily status you should change it otherwise we can wait for SB 2.1.4

pvliss commented:
@pascalgrimaud Maybe we should wait for SB 2.1.4 to be released before merging/changing this to see if the error is solved as suggested by @DanielFran https://github.com/jhipster/generator-jhipster/issues/9513#issuecomment-478372606?

pvliss commented:
Also in favor of using SDKMan nonetheless but also curious to see if SB 2.1.4 does indeed also fix the problem :smile:

pascalgrimaud commented:
let's wait tomorrow
I'm putting hold on tag on this PR

PierreBesson commented:
@pvliss I don't understand why SDKMan jdk fixes the problem. It should not be too different from the jdk provided by debian.

Oh and it seems that sdkman installs the Zulu JDK ! Personally I'm against using it.

pvliss commented:
@PierreBesson To be honest I am not sure either but as I was testing locally with SDKMan I could not reproduce the problem and thought I should try. I also had a look last night at the bug fix for SB 2.1.4 after @DanielFran's comment and it seems weird tom me too that it works with SDKMan provided JDK but not the one supplied by Ubuntu, hence why I said we should better wait for the fix. I will check again a bit later and try to understand where's the difference/problem.

Regarding JDK, SDKMan supports several dists(openjdk, zuul, oracle, correto, etc.). I thought I've set it up to use the OpenJDK one. Will check again

PierreBesson commented:
Yes sdkman does let you install many options. Oh and I see that you use the `java-open` version so that should not be the zulu jvm.
I'm personnally using sdkman for my dev machine as well. It can also be helpful to test different version like jdk12 that should already work.

So I'm changing my mind. Let's merge your PR. Then could you please add a build for jdk12 to the daily builds project. Thanks a lot.

pvliss commented:
Yes, I believe it correctly installs open JDK. I have added a `java --version` call and if you look at the logs you will see: `OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+9, mixed mode)` so I guess that is right.

Glad you changed your mind. BTW, I have also changed the docker file to use SDK Man as well.

As I said also in favor of using SDKMan for the following reasons:
1. Been using for a 1-2 years now and never had an issue with it.
2. It provides many options to choose from and it will be very easy for the end users who use docker to change to whatever they prefer or even experiment with newer versions

It can also be used to make CI a bit easier to maintain. Take a look at this example: http://eed3si9n.com/all-your-jdks-on-travis-ci-using-sdkman

pvliss commented:
So I also tested with OpenJDK 11.0.1(downloaded from the [archive](https://jdk.java.net/archive/)) in my Windows laptop and everything works properly as well. Also tried finding something related in the Open JDK tracker for [11.0.2](https://bugs.openjdk.java.net/issues/?jql=project+%3D+JDK+AND+fixVersion+%3D+11.0.2) but again not sure if something relates to this.
Maybe it is indeed related to the dist of Ubuntu as it seems to be kind of messed up(see [this SO thread](https://askubuntu.com/questions/1037646/why-is-openjdk-10-packaged-as-openjdk-11) which also contains an alternative solution to using OpenJDK 11.0.2 as well)? Not sure either. Hence I give up.  Better spend time elsewhere.

PierreBesson commented:
Oh this is actually a smart trick on ubuntu's part as now they can upgrade openjdk to version 11 for bionic otherwise they would have been stuck with 10. This is because bionic is an LTS release sl they are not supposed to upgrade their packages to new major versions for the duration of the support.

Anyway the PR looks good to me. I will let @pascalgrimaud merge this tommorow.

pascalgrimaud commented on .travis.yml:
```diff
@@ -70,18 +70,19 @@ before_install:
-    - sudo apt-get install -y openjdk-11-jdk
-    - sudo update-java-alternatives -s java-1.11.0-openjdk-amd64
-    - java -version
+    - curl -s get.sdkman.io | bash
```
I was going to ask you why install 2 times jdk11 here, but it was already the case in current master! And it was my fault :)
Could you remove this part as it is already done above when JHI_JDK is 11

pvliss replied:
Hello @pascalgrimaud. Sorry for the delayed response. Fixed

pascalgrimaud commented:
thanks for your hard work @pvliss

pvliss commented:
Glad to be of help
